### DISCUSSION TITLE
How to update response cache when using useMutation

### DISCUSSION THREAD

**Discussion by yokoyama-takuya** [TanStack/query#5850] [Q&A]
Thanks for awesome library!
I want to update mutation's cache without refetching, is that possible? like `queryClient.setQueryData`.
https://tanstack.com/query/latest/docs/react/reference/QueryClient#queryclientsetquerydata

I know can use any HTTP request method  by `useQuery's queryFn` and  ` useMutation's mutationFn `  .
Even if POST method's API, I should use `useQuery`?
The API is only search, but using POST method  becase have many request parametes.

**Comment by TkDodo** [comment]
I think you want this: https://tanstack.com/query/v4/docs/react/guides/updates-from-mutation-responses

**Comment by yokoyama-takuya** [comment]
@TkDodo Thank you for your reply.  I want to update mutation's cache.

```.ts
  const mutation = useMutation<{ list: { data: string }[] }>({})
  if (mutation.data?.list) {
    // I want to change this `mutation.data.list`
  }
```

**Reply by TkDodo** [reply]
transform the response inside the `mutationFn`, before returning it

**Comment by PinkaminaDianePie** [comment]
also have this problem. it looks like mutations don't expose any `setData` method :(

**Reply by TkDodo** [reply]
https://github.com/TanStack/query/discussions/5850#discussioncomment-6780129

**Reply by PinkaminaDianePie** [reply]
> [#5850 (reply in thread)](https://github.com/TanStack/query/discussions/5850#discussioncomment-6780129)

Ah, ok, i don't have exactly the same problem. I need to update response cache not from the mutation response, but from the different place, so i need `setData`, not `mutationFn`

**Reply by TkDodo** [reply]
First you say you have the same problem, but then it's not the exact same problem ...

"I need ..." is not very helpful. You're asking a classic XY question, which, sorry to say, is wasting my time.

If you want my help, please read this and come back with a question I can answer. Help me help you.

https://tkdodo.eu/blog/how-can-i

**Reply by PinkaminaDianePie** [reply]
If I'd ask as this blog post suggests, it will be an even bigger waste of time :D 
Original problem: "How do I implement a normalized cache with a react query?"
The problem is, I don't need an answer to that problem. I know how to normalize the data, I know how to propagate it back to all the active queries, it already works, and it works fine for my particular use case. The problem is that it works only for queries. The query has `setData` method which I can use to update it when it's data is updated by another query/mutation. So the only thing I'm missing is `how to update mutation result? it doesn't expose setData, so I don't know how can i do it`. That's what I've asked: `I need to update response cache not from the mutation response, but from the different place`. I don't need you to explain me how cache works in general, why react query decided not to go with normalized caches in the first place, and so on. I can figure it out. I just need to know how to update the mutation result from the other place, that's it.

**Reply by TkDodo** [reply]
Believe it or not, this context helps.

> I need to update response cache not from the mutation response, but from the different place

The bigger underlying issue is that mutations aren't really a global state manager like queries. With a query, you can call `useQuery` in multiple components and it will read the same data from the same cache entry.

If you call `useMutation` twice in different components, you will not get the same `data`. If you call `.mutate()` in both, they will each have their own cache instance. You can see that in the devtools.

Mutations aren't "meant" for distributing data throughout your application, so really, setting data on a mutation shouldn't be something that you want.

Ideally, a mutation updates a resource on the server, and then either updates the corresponding query or invalidates it. Then, you'll use the result of the query to access data - but not the `data` returned from the mutation.

**Reply by PinkaminaDianePie** [reply]
> Ideally

That's the key word here :) I follow this pattern in the new code, but I have some legacy stuff, like where the screen shows the result of the mutation, and this result can be changed later on by websocket-based subscription. I've thought I could mimic this behavior by updating the mutations from the caching logic to avoid any refactorings. But with those limitations in place, it's probably just easier to rewrite such places to redirect right after mutation to a page with a query.

**Reply by TkDodo** [reply]
yeah I just checked there is no API that does it. You can't even try to find the mutation in the mutationCache and then call `setData` because it doesn't exist.

You can call `setOptions` and set a new `mutationFn` that synchronously writes, and then call `.execute()` to execute it but please don't do that 😂

**Reply by PinkaminaDianePie** [reply]
> You can call setOptions and set a new mutationFn that synchronously writes, and then call .execute() to execute it but please don't do that 😂

That feels too hacky, i'd rather rewrite my code to stop relying on mutation results that much :D

**Reply by TkDodo** [reply]
good call :)
